2020/5/binary_seats.rb

```diff
@@ -0,0 +1,51 @@
+# FBFBBFFRLR
+# 0-127
+# F: 0-63
+# B; 32-63
+# F: 32-47
+# ...
+def binary_move(start_i, end_i, char, left_marker)
+  middle = start_i + (end_i - start_i) / 2
+  char == left_marker ? [start_i, middle] : [middle + 1, end_i]
+end
+
+def binary_partition(seat)
+  row = seat[0..6].chars.reduce([0, 127]) do |(start_i, end_i), char|
+    binary_move(start_i, end_i, char, 'F')
+  end.first
+  column = seat[7..9].chars.reduce([0, 7]) do |(start_i, end_i), char|
+    binary_move(start_i, end_i, char, 'L')
+  end.first
+  row * 8 + column
+end
+
+# @param seat [String] encoded seat number like 'FBFFBBFRLR'
+# @return [Integer] decoded seat number
+# decodes row by converting encoded row to binary number
+# does same thing for column
+def binary_partition2(seat)
```

Comment:
Realized this after seeing some other people's code, the encoded seat number is a binary number. Moving to the back / right means you move `2^i` rows / columns.
Example calculation for a row:
```
FBFFBBF, 0
0100110, 0
|           0 + 0 * 2^6 = 0
0100110, 0
 |          0 + 1 * 2^5 = 32
0100110, 32
  |         32 + 0 * 2^4 = 32
0100110, 32
   |        32 + 0 * 2^3 = 32
0100110, 32
    |       32 + 1 * 2^2 = 36
0100110, 36
     |      40 + 1 * 2^1 = 38
0100110, 38
      |     44 + 0 * 2^2 = 38
```